test/perf.jl

```diff
@@ -0,0 +1,82 @@
+using BenchmarkTools
+using Setfield
+using Test
+using InteractiveUtils
+
+struct AB{A,B}
+    a::A
+    b::B
+end
+
+function lens_set_a(obj, val)
+    @set obj.a = val
+end
+
+function hand_set_a(obj, val)
+    AB(val, obj.b)
+end
+
+function lens_set_ab(obj, val)
+    @set obj.a.b = val
+end
+
+function hand_set_ab(obj, val)
+    a = AB(obj.a.a, val)
+    AB(a, obj.b)
+end
+
+function lens_set_a_and_b(obj, val)
+    o1 = @set obj.a = val
+    o2 = @set o1.b = val
+end
+
+function hand_set_a_and_b(obj, val)
+    AB(val, val)
+end
+
+function isgetfield(ex)
+    ex.head == :call &&
+    ex.args[1] == GlobalRef(Base, :getfield)
+end
+
+function isnew(ex)
+    ex.head == :new
+end
+isreturn(ex) = ex.head == :return
+
+function is_fast_code(info::Core.CodeInfo)
+    for ex in info.code
+        isgetfield(ex) ||
+        isnew(ex) ||
+        isreturn(ex) ||
+        return false
+    end
+    true
+end
+
+@testset "benchmark" begin
+    obj = AB(AB(1,2), :b)
+    val = (1,2)
+    for (f_lens, f_hand) in [
+                             (lens_set_a, hand_set_a),
+                             (lens_set_ab, hand_set_ab),
+                             (lens_set_a_and_b, hand_set_a_and_b)
+                            ]
+        b_lens = @benchmark $f_lens($obj, $val)
+        b_hand = @benchmark $f_hand($obj, $val)
+
+        # actually they should be equal
+        # but there is too much noise
+        @test minimum(b_lens).time < 2*minimum(b_hand).time
```

Comment:
Maybe you've already tried, but I think another way to do this to use `judge` and make the test pass if the lens is either marked as faster or invariant.  IIUC `judge` would take the fluctuations in the samples into account so it could be more robust.

Reply:
Thats a good idea. I will try it. Also need to study `BenchmarkTools` more closly, there might be other useful functionality.